libyul/optimiser/LabelIDDispenser.cpp

```diff
@@ -0,0 +1,163 @@
+)
+{
+	return isRestrictedIdentifier(_dialect, _label) || _reservedLabels.contains(_label);
+}
+}
+
+LabelIDDispenser::LabelIDDispenser(ASTLabelRegistry const& _labels, std::set<std::string> const& _reservedLabels):
+	m_labels(_labels),
+	m_reservedLabels(_reservedLabels.begin(), _reservedLabels.end()),
+	m_offset(_labels.maxID() + 1)
+{
+	m_reservedLabels += std::set{""};
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newID(LabelID const _parent)
+{
+	m_idToLabelMapping.push_back(resolveParentLabelID(_parent));
+	return m_idToLabelMapping.size() - 1 + m_offset;
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::newGhost()
+{
+	m_idToLabelMapping.push_back(ASTLabelRegistry::ghostLabelIndex());
+	return m_idToLabelMapping.size() - 1 + m_offset;
+}
+
+LabelIDDispenser::LabelID LabelIDDispenser::resolveParentLabelID(LabelID _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_offset, "ID exceeds bounds.");
+	// bigger than offset means, that the input label id that was spawned by this dispenser
+	if (_id >= m_offset)
+		_id = m_idToLabelMapping[_id - m_offset];
+	yulAssert(
+		_id < m_offset && !m_labels.unused(_id),
+		"We can have at most one level of indirection and the derived-from label cannot be unused."
+	);
+	return _id;
+}
+
+bool LabelIDDispenser::ghost(LabelID const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size() + m_offset, "ID exceeds bounds.");
+	if (_id >= m_offset)
+		return m_idToLabelMapping[_id - m_offset] == ASTLabelRegistry::ghostLabelIndex();
+
+	return m_labels.ghost(_id);
+}
+
+
+ASTLabelRegistry LabelIDDispenser::generateNewLabels(Block const&, Dialect const& _dialect) const
+{
+	// this can be replaced by the actually used ids in the provided block once the AST uses ids instead of YulString
+	std::set<LabelID> usedIDs = ranges::views::iota(static_cast<size_t>(0), m_idToLabelMapping.size() + m_offset) | ranges::to<std::set<LabelID>>;
+
+	if (usedIDs.empty())
+		return {};
+
+	auto const& originalLabels = m_labels.labels();
+
+	std::vector<uint8_t> reusedLabels (originalLabels.size());
+	// this means that everything that is derived from empty needs to be generated
+	reusedLabels[0] = true;
+
+	// start with the empty label
+	std::vector<std::string> labels{""};
+	labels.reserve(originalLabels.size()+1);
+
+	// usedIDs contains at least one element at this point
+	std::vector<size_t> idToLabelMap(*std::prev(usedIDs.end()) + 1, 0);
+
+	std::set<std::string, std::less<>> alreadyDefinedLabels = m_reservedLabels;
+
+	// we record which labels have to be newly generated, some we can just take over from the existing registry
+	std::vector<LabelID> toGenerate;
+	for (auto const& id: usedIDs)
+	{
+		if (ghost(id))
+		{
+			idToLabelMap[id] = ASTLabelRegistry::ghostLabelIndex();
+			continue;
+		}
+
+		auto const parentLabelID = resolveParentLabelID(id);
+
+		auto const originalLabelIndex = m_labels.idToLabelIndex(parentLabelID);
+		auto const& originalLabel = originalLabels[originalLabelIndex];
```

Comment:
This is a string view, not a string. Is there any point in taking a reference to it?
```suggestion
		std::string_view const originalLabel = originalLabels[originalLabelIndex];
```

BTW, this yet another good example of how `auto` obfuscates things that would have been obvious otherwise.

Reply:
it is a string :) case in point, though!